zcash_client_sqlite/src/wallet/init/migrations/tx_retrieval_queue.rs

```diff
@@ -0,0 +1,87 @@
+//! A migration to add the `tx_retrieval_queue` table to the database.
+
+use rusqlite::{named_params, Transaction};
+use schemer_rusqlite::RusqliteMigration;
+use std::collections::HashSet;
+use uuid::Uuid;
+use zcash_primitives::transaction::builder::DEFAULT_TX_EXPIRY_DELTA;
+
+use crate::wallet::init::WalletMigrationError;
+
+use super::utxos_to_txos;
+
+pub(super) const MIGRATION_ID: Uuid = Uuid::from_u128(0xfec02b61_3988_4b4f_9699_98977fac9e7f);
+
+pub(crate) struct Migration;
+
+impl schemer::Migration for Migration {
+    fn id(&self) -> Uuid {
+        MIGRATION_ID
+    }
+
+    fn dependencies(&self) -> HashSet<Uuid> {
+        [utxos_to_txos::MIGRATION_ID].into_iter().collect()
+    }
+
+    fn description(&self) -> &'static str {
+        "Adds a table for tracking transactions to be downloaded for transparent output and/or memo retrieval."
+    }
+}
+
+impl RusqliteMigration for Migration {
+    type Error = WalletMigrationError;
+
+    fn up(&self, transaction: &Transaction) -> Result<(), WalletMigrationError> {
+        transaction.execute_batch(
+            "CREATE TABLE tx_retrieval_queue (
+                txid BLOB NOT NULL UNIQUE,
+                query_type INTEGER NOT NULL,
+                dependent_transaction_id INTEGER,
+                FOREIGN KEY (dependent_transaction_id) REFERENCES transactions(id_tx)
+            );
+
+            ALTER TABLE transactions ADD COLUMN target_height INTEGER;
+
+            CREATE TABLE transparent_spend_search_queue (
+                address TEXT NOT NULL,
+                transaction_id INTEGER NOT NULL,
+                output_index INTEGER NOT NULL,
+                FOREIGN KEY (transaction_id) REFERENCES transactions(id_tx),
+                CONSTRAINT value_received_height UNIQUE (transaction_id, output_index)
+            );",
+        )?;
+
+        // Add estimated target height information for each transaction we know to
+        // have been created by the wallet; transactions that were discovered via
+        // chain scanning will have their `created` field set to `NULL`.
```

Comment:
What is the consequence of this being incorrect, e.g. when a non-default expiry delta was used?

Reply:
This is a heuristic that is essentially targeted to catch _only_ ZIP-320 second transactions in wallet recovery:
* In the course of normal wallet operation, the wallet will have an entry in the `transactions` table already for that transaction, so we don't need to discover it.
* In the course of wallet recovery, there will ordinarily be change belonging to the wallet that can be used to further explore the transaction graph. The remaining case where history recovery will fail is if the full value of a transparent UTXO is spent outside of the expiry delta. To fix this, we still need to add functionality to compare "assumed UTXOs" to "actual UTXOs" but in order to do so correctly we need a lightwalletd API that looks up transactions by their input `OutPoint` values. When we have such an API, we can then fix this case properly. I think that there's a follow-up possible here that compares the actual UTXO set to the assumed UTXO set, and I've actually made a couple of attempts at this, but I don't think I have figured out a really good approach yet, so I would prefer not to block this PR on it.